### PR TITLE
Fix DevEncoded commands on Python 3

### PR DESCRIPTION
**Pull request by ajoubertza**
Work in progress.
- [x] Added unit test to ensure that DevEncoded (aka bytearray) commands work the same as the attributes.
- [x] Fix implementation so that tests pass. 

Fixes:  issue #280

### CODE REVIEW

File: ext/server/command.cpp [hunk 1]
```diff
@@ -222,9 +222,22 @@ void extract_scalar<Tango::DEV_ENCODED>(const CORBA::Any &any, boost::python::ob
         throw_bad_type(Tango::CmdArgTypeName[Tango::DEV_ENCODED]);
     
     bopy::str encoded_format(data[0].encoded_format);
+
+#if PY_MAJOR_VERSION >= 3
+    boost::python::object encoded_data(
+        boost::python::handle<>(
+            PyMemoryView_FromMemory(
```
Comment (ajoubertza) [inline ext/server/command.cpp]: Thanks for working on this.

- As mentioned in the issue thread, we have similar code in `to_py.h`, so I wonder about the difference.
- Second point:  that code doesn't do a Python 3 check, so wonder if we need it here?
      
https://github.com/tango-controls/pytango/blob/19e4c5daf5916384b030b4fb1e373d72d1c0c0e7/ext/to_py.h#L20-L31

**Comment by sdebionne** [comment]
@ajoubertza The latest pipeline was set to build with tango 9.2.5a in `.travis.yml`. Now the test fail with a link error:

```
E   ImportError: /home/travis/build/tango-controls/pytango/tango/_tango.so: undefined symbol: _ZN5Tango10DeviceDatarsERPKNS_18DevVarBooleanArrayE
```

which let me think that the CI either used a cached build of pytango to run the tests (scarry) or that the build and test environments are different. Any clue?

**Comment by ajoubertza** [comment]
> which let me think that the CI either used a cached build of pytango to run the tests (scarry) or that the build and test environments are different. Any clue?

Yes, the builds do use caching, so as not to rebuild the tango wrapper extension unnecessarily.  The build looks at the source files, but I guess it doesn't know if the conda environment changed. 

You could try modifying the `.travis.yml` file.  Maybe removing the `cache:` section?  Or add a
`rm -rf cached_ext`  just before `mkdir -p cached_ext`.

Do you really want to test against 9.2.5a?

File: ext/server/command.cpp [hunk 2]
```diff
@@ -222,9 +222,21 @@ void extract_scalar<Tango::DEV_ENCODED>(const CORBA::Any &any, boost::python::ob
         throw_bad_type(Tango::CmdArgTypeName[Tango::DEV_ENCODED]);
     
     bopy::str encoded_format(data[0].encoded_format);
+
+#if PY_MAJOR_VERSION >= 3
+    boost::python::object encoded_data(
```
Comment (tiagocoutinho) [inline ext/server/command.cpp:225]: couldn't we use the new code for both python 2 and python 3?

**Review by tiagocoutinho — CHANGES_REQUESTED** [review]

**Comment by ajoubertza** [comment]
@sdebionne I tried your changes in my environment (Docker container), with TANGO 9.3.2alpha.1 conda package.  They worked fine.

Even after changes in c131db3, the extension was not recompiled by Travis.  However, I see an option to delete the cache on Travis today (didn't see it yesterday).  I deleted the cache for this PR but that didn't help either!

Finally read the [Travis docs](https://docs.travis-ci.com/user/caching/#fetching-and-storing-caches).  Interestingly, if the PR branch doesn't have a cache yet, it takes one from the default branch.

I notice a bug in the `.travis.yml` file - when checking if the extenstion source files have changed, the diff was not recursive.  Trying that next.

**Comment by ajoubertza** [comment]
That worked.  Tests passing at last.

File: ext/server/command.cpp (hunk 2 above)
Comment (sdebionne) [reply ext/server/command.cpp:225]: I  guess so, it's in c131db3.

**Comment by sdebionne** [comment]
I had a day off yesterday. Glad to hear that travis config was buggy and fixed!

> Do you really want to test against 9.2.5a?

Nope, it was more a desperate attempt to understand what was wrong with the pipeline and limit the differences between my local build env and the CI. Thanks for reverting this.

Everyting LGTM!

**Review by tiagocoutinho — APPROVED** [review]
LGTM

**Pull request merged**